api/client/proxy/client.go

```diff
@@ -115,6 +118,22 @@ func (c *ClientConfig) CheckAndSetDefaults() error {
 			// not to send the client certificate by looking at certificate request.
 			if len(tlsCfg.Certificates) > 0 {
 				cert := tlsCfg.Certificates[0]
+
+				// When a hardware key is used to store the private key, the user may fail to provide
+				// a PIN or touch before a gRPC dial timeout occurs.
+				// The resulting "dial timeout" error is generic and doesn't indicate an issue with the
+				// hardware key itself (since YubiKey is treated like any other key).
+				// To avoid this, we perform a "warm-up" call to the key, ensuring it is ready
+				// before initiating the gRPC dial.
+				// This approach works because the connection is cached for a few seconds,
+				// allowing subsequent calls without requiring additional user action.
+				if priv, ok := cert.PrivateKey.(*keys.YubiKeyPrivateKey); ok {
+					b := make([]byte, 256)
+					_, err := priv.Sign(rand.Reader, b, crypto.SHA256)
```

Comment:
I just realized that with this warm up code added, I'm not able to interrupt the program with Ctrl-C. I guess it's because prompting for password/touch blocks the program execution? I'm not sure where I should look at.

Reply:
I think we handle ctrl-c by just cancelling a context and hoping everything exits quickly after that. You could try my suggestion here with passing down a context https://github.com/gravitational/teleport/pull/47091#discussion_r1784902191

Reply:
Fortunately, there is a context in `NewClient`, so I passed it down to `CheckAndSetDefaults` and then to `WarmupHardwareKey`.
After some debugging, I realized that the problem with handling Ctrl-C was caused by the `singer.Sign` function that blocked the program, waiting for the user to provide a PIN or touch. My workaround for this was to wrap it in a goroutine and abandon the result when the context is canceled. I believe it is a standard way of dealing with such problems in Go.